Fix boolean custom fields in PostgreSQL even more

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19601
## Description

Turns out that even more code assumed that `true == "1"`...
For some occurences I replaced legacy methods by calls to `CustomValue.typed_value`, for others I just fixed the offending code immediately, keeping the duplication in place.

This is sad, but true...

NobodysNightmare committed:
replace occurences of cast_value with typed_value

NobodysNightmare committed:
enhance PSQL compatibility of cast_value

In PSQL a valid boolean can also be represented using other value than
'0' and '1'

This led to cast_value sometimes returning false, even though it was supposed to return true...

NobodysNightmare committed:
refactor initializer to use named parameters

NobodysNightmare committed:
fix bool conversion in custom field format

this should probably be merged with occurences of:

- CustomValue.typed_value
- CustomField.cast_value

I am not yet sure how to best avoid all that duplication...

NobodysNightmare committed:
ensure that format always returns a string

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-fix-boolean-in-psql-even-more.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-fix-boolean-in-psql-even-more.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented:
@NobodysNightmare specs fail for real!

NobodysNightmare committed:
fix expectations

- when the value is '' we should expect '' and not nil (IMO the old implementation was flawed)
- do not care whether find_by_id is called with string or int

NobodysNightmare committed:
refactor specs

- use RSpec features to highlight common behaviour between specs
- do not mock methods on the subject

floriank commented:
@myabc are you still on the review or are you self-assigned by accident?
